comparative_benchmark/benchmark_lib.py

```diff
@@ -0,0 +1,223 @@
+         warmup_iterations: int, iterations: int,
+         input_npys: Sequence[pathlib.Path],
+         expect_npys: Sequence[pathlib.Path],
+         benchmark_function: Callable) -> utils.BenchmarkResult:
+  model = benchmark.model
+  input_data = benchmark.input_data.artifacts[
+      def_types.ModelTestDataFormat.NUMPY_TENSORS]
+  expected_output = benchmark.expected_output.artifacts[
+      def_types.ModelTestDataFormat.NUMPY_TENSORS]
+
+  data_type = model.model_parameters["data_type"]
+  batch_size = model.model_parameters["batch_size"]
+  input_dims = input_data.data_parameters["tensor_dimensions"]
+  output_dims = expected_output.data_parameters["tensor_dimensions"]
+  benchmark_definition = {
+      "benchmark_id": benchmark.id,
+      "benchmark_name": benchmark.name,
+      "framework": str(model.model_impl.framework_type),
+      "data_type": data_type,
+      "batch_size": batch_size,
+      "inputs": input_dims,
+      "outputs": output_dims,
+      "compiler": "xla",
+      "device": benchmark.target_device.name,
+      "tags": model.model_impl.tags + model.tags,
+  }
+
+  print(f"\n\n--- {benchmark.name} ---")
+
+  # Retrieve framework-level benchmarks.
+  with multiprocessing.Manager() as manager:
+    shared_dict = manager.dict()
+
+    backend = benchmark.target_device.accelerator_type
+    kwargs: Dict[str, Any] = dict(
+        model=model,
+        input_npys=list(input_npys),
+        expect_npys=list(expect_npys),
+        verify_params=expected_output.verify_parameters,
+        warmup_iterations=warmup_iterations,
+        benchmark_iterations=iterations,
+        backend=backend,
+        shared_dict=shared_dict,
+    )
+    if run_in_process:
+      benchmark_function(**kwargs)
+    else:
+      p = multiprocessing.Process(target=benchmark_function, kwargs=kwargs)
+      p.start()
+      p.join()
+
+    framework_metrics = dict(shared_dict)
+
+  return utils.BenchmarkResult(
+      definition=benchmark_definition,
+      metrics={
+          "framework_level": framework_metrics,
+      },
+  )
+
+
+def _download_artifacts(benchmarks: Sequence[def_types.BenchmarkCase],
+                        root_dir: pathlib.Path,
+                        verbose: bool = False):
+  """Download benchmark artifacts."""
+
+  download_list = []
+  for benchmark in benchmarks:
+    input_artifact = benchmark.input_data.artifacts[
+        def_types.ModelTestDataFormat.NUMPY_TENSORS]
+    input_path = root_dir / benchmark.model.name / "input_npy.tgz"
+    download_list.append((input_artifact.source_url, input_path))
+
+    expect_artifact = benchmark.expected_output.artifacts[
+        def_types.ModelTestDataFormat.NUMPY_TENSORS]
+    expect_path = root_dir / benchmark.model.name / "output_npy.tgz"
+    download_list.append((expect_artifact.source_url, expect_path))
+
+  utils.download_files(download_list, verbose=verbose)
+
+
+def parse_arguments() -> argparse.ArgumentParser:
```

Comment:
Maybe take the argument parser as a parameter and configure, then return it. This way different tools can set their own descriptions.

And should this be called `configure_parser`?

The example can be found here

https://github.com/openxla/openxla-benchmark/blob/7fe74c99bfe4568d9a0f48091635a0dda59695c3/devtools/db_import/db_import/batch_import.py#L20-L29

Reply:
Thanks, that looks a lot better.